File: sample/src/main/java/com/readdle/swiftjava/sample/UInt64Test.kt
package com.readdle.swiftjava.sample

import com.readdle.codegen.anotation.*
import java.lang.annotation.Native

private const val MIN_VALUE: Long = 0
private const val MAX_VALUE: Long = -1

@SwiftValue
enum class UInt64Enum(@Unsigned val rawValue: Long) {

    ONE(0), TWO(1), THREE(2);

    companion object {

        private val values = HashMap<Long, UInt64Enum>()

        @JvmStatic
        fun valueOf(rawValue: Long): UInt64Enum {
            return values[rawValue]!!
        }

        init {
            enumValues<UInt64Enum>().forEach {
                values[it.rawValue] = it
            }
        }
    }

}

@SwiftValue
data class UInt64OptionsSet(@Unsigned val rawValue: Long = 0) {
    companion object {
        @JvmStatic
        val  one = UInt64OptionsSet(1)
        @JvmStatic
        val  two = UInt64OptionsSet(2)
        @JvmStatic
        val  three = UInt64OptionsSet(4)
    }
}

@SwiftValue
data class UInt64TestStruct(@Unsigned var zero: Long = 0,
                            @Unsigned var max: Long = MAX_VALUE,
                            @Unsigned var min: Long = MIN_VALUE,
                            @Unsigned var optional: Long? = 0,
                            @Unsigned var optionalNil: Long? = null)

@SwiftReference
class UInt64Test private constructor() {

    @SwiftDelegate(protocols = ["UInt64TestParamProtocol"])
    interface UInt64ParamProtocol {
        @SwiftCallbackFunc
        fun testParam(@Unsigned param: Long): Boolean
    }

    @SwiftDelegate(protocols = ["UInt64TestReturnTypeProtocol"])
    interface UInt64ReturnTypeProtocol {
        @SwiftCallbackFunc @Unsigned
        fun testReturnType(): Long
    }

    @SwiftDelegate(protocols = ["UInt64TestOptionalParamProtocol"])
    interface UInt64OptionalParamProtocol {
        @SwiftCallbackFunc
        fun testOptionalParam(@Unsigned param: Long?): Boolean
    }

    @SwiftDelegate(protocols = ["UInt64TestOptionalReturnTypeProtocol"])
    interface UInt64OptionalReturnTypeProtocol {
        @SwiftCallbackFunc @Unsigned
        fun testOptionalReturnType(): Long?
    }

    @SwiftBlock("(UInt64) -> UInt64")
    interface UInt64Block {
        @Unsigned
        fun call(@Unsigned value: Long): Long
    }

    @SwiftBlock("(UInt64?) -> UInt64?")
    interface OptionalUInt64Block {
        @Unsigned
        fun call(@Unsigned value: Long?): Long?
    }

    companion object {
        @JvmStatic @Unsigned
        external fun testZero(): Long

        @JvmStatic @Unsigned
        external fun testMin(): Long

        @JvmStatic @Unsigned
        external fun testMax(): Long

        @JvmStatic
        external fun testParam(@Unsigned param: Long): Boolean

        @JvmStatic @Unsigned
        external fun testReturnType(): Long

        @JvmStatic
        external fun testOptionalParam(@Unsigned param: Long?): Boolean

        @JvmStatic @Unsigned
        external fun testOptionalReturnType(): Long?

        @JvmStatic
        external fun testProtocolParam(callback: UInt64ParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolReturnType(callback: UInt64ReturnTypeProtocol): Long

        @JvmStatic
        external fun testProtocolOptionalParam(callback: UInt64OptionalParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolOptionalReturnType(callback: UInt64OptionalReturnTypeProtocol): Long?

        @JvmStatic
        external fun testEncode():  UInt64TestStruct

        @JvmStatic
        external fun testDecode(value: UInt64TestStruct): Boolean

        @JvmStatic
        external fun testEnumEncode(@Unsigned rawValue: Long) : UInt64Enum

        @JvmStatic @Unsigned
        external fun testEnumDecode(enum: UInt64Enum) : Long

        @JvmStatic
        external fun testOptionSetEncode(@Unsigned rawValue: Long) : UInt64OptionsSet

        @JvmStatic @Unsigned
        external fun testOptionSetDecode(enum: UInt64OptionsSet) : Long
        
        @JvmStatic
        external fun testBlock(@SwiftBlock block: UInt64Block): Boolean

        @JvmStatic
        external fun testOptionalBlock(@SwiftBlock block: OptionalUInt64Block): Boolean
    }

    @Native
    var nativePointer: Long = 0

    external fun release()

}

File: sample/src/main/swift/Sources/SampleAppCore/UInt64Test.swift
//
// Created by Andrew on 2/4/18.
//

import Foundation

public enum UInt64Enum: UInt64, Codable {
    case one
    case two
    case three
}

public struct UInt64OptionsSet: OptionSet, Codable {

    public let rawValue: UInt64

    public init(rawValue: UInt64) {
        self.rawValue = rawValue
    }

    static let one = UInt64OptionsSet(rawValue: 1 << 0)
    static let two = UInt64OptionsSet(rawValue: 1 << 1)
    static let three = UInt64OptionsSet(rawValue: 1 << 2)
}

public struct UInt64TestStruct: Codable, Hashable {
    public var zero: UInt64 = UInt64.zero
    public var max: UInt64 = UInt64.max
    public var min: UInt64 = UInt64.min
    public var optional: UInt64? = UInt64.zero
    public var optionalNil: UInt64? = nil
}

public protocol UInt64TestParamProtocol {
    func testParam(_ param: UInt64) -> Bool
}

public protocol UInt64TestReturnTypeProtocol {
    func testReturnType() -> UInt64
}

public protocol UInt64TestOptionalParamProtocol {
    func testOptionalParam(_ param: UInt64?) -> Bool
}

public protocol UInt64TestOptionalReturnTypeProtocol {
    func testOptionalReturnType() -> UInt64?
}

public typealias UInt64Block = (_ value: UInt64) -> UInt64
public typealias OptionalUInt64Block = (_ value: UInt64?) -> UInt64?

public class UInt64Test {

    public static func testZero() -> UInt64 {
        return 0
    }

    public static func testMin() -> UInt64 {
        return UInt64.min
    }

    public static func testMax() -> UInt64 {
        return UInt64.max
    }

    public static func testParam(_ param: UInt64) -> Bool {
        return param == UInt64.max
    }

    public static func testReturnType() -> UInt64 {
        return UInt64.max
    }

    public static func testOptionalParam(_ param: UInt64?) -> Bool {
        return param == UInt64.max
    }

    public static func testOptionalReturnType() -> UInt64? {
        return UInt64.max
    }

    public static func testProtocolParam(_ callback: UInt64TestParamProtocol) -> Bool {
        return callback.testParam(UInt64.max)
    }

    public static func testProtocolReturnType(_ callback: UInt64TestReturnTypeProtocol) -> UInt64 {
        return callback.testReturnType()
    }

    public static func testProtocolOptionalParam(_ callback: UInt64TestOptionalParamProtocol) -> Bool {
        return callback.testOptionalParam(UInt64.max)
    }

    public static func testProtocolOptionalReturnType(_ callback: UInt64TestOptionalReturnTypeProtocol) -> UInt64? {
        return callback.testOptionalReturnType()
    }

    public static func testEncode() ->  UInt64TestStruct {
        return UInt64TestStruct()
    }

    public static func testDecode(_ value: UInt64TestStruct) -> Bool {
        return value == UInt64TestStruct()
    }

    public static func testEnumEncode(_ rawValue: UInt64) -> UInt64Enum {
        switch rawValue {
        case UInt64Enum.one.rawValue: return UInt64Enum.one
        case UInt64Enum.two.rawValue: return UInt64Enum.two
        case UInt64Enum.three.rawValue: return UInt64Enum.three
        default: fatalError("Can't find enum with rawValue \(rawValue)")
        }
    }

    public static func testEnumDecode(_ enumValue: UInt64Enum) -> UInt64 {
        return enumValue.rawValue
    }

    public static func testOptionSetEncode(_ rawValue: UInt64) -> UInt64OptionsSet {
        return UInt64OptionsSet(rawValue: rawValue)
    }

    public static func testOptionSetDecode(_ optionSet: UInt64OptionsSet) -> UInt64 {
        return optionSet.rawValue
    }

    public static func testBlock(_ block: UInt64Block) -> Bool {
        let value = block(0)
        return value == 0
    }

    public static func testOptionalBlock(_ block: OptionalUInt64Block) -> Bool {
        let value = block(nil)
        return value == nil
    }

}